### PR TITLE
Add guide on unit testing

### PR DESCRIPTION
**Pull request by qunabu**
#### Description of what you did:

Improved version of https://github.com/strapi/strapi/pull/6324 

Documentation update on unit testing based on my medium post https://medium.com/qunabu-interactive/strapi-jest-testing-with-gitlab-ci-82ffe4c5715a.

Sample repository to that includes the examples from the guide https://github.com/qunabu/strapi-unit-test-example

### CODE REVIEW

File: docs/v3.x/guides/unit-testing.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+Test framework must have a clean empty environment to perform valid test and also not to interfere with current database.
+
+Once `jest` is running it uses the `test` [enviroment](../concepts/configurations.md#environments) (switching `NODE_ENV` to `test`)
+so we need to create a special environment setting for this purpose.
+Copy contents of `./config/environments/development` into `./config/environments/test` and add some changes.
+Edit `./config/environments/test/database.json` and add value `"filename": ".tmp/test.db"` - the reason of that is that we want to have a separate sqlite database for tests, so our test will not touch real data.
+This file will be temporary, each time test is finished, we will remove that file that every time tests are run on the clean database.
+The whole file will look like this:
+
+**Path —** `./config/environments/test/database.json`
```
Comment (alexandrebodin) [inline docs/v3.x/guides/unit-testing.md:73]: You will need to change the paths here :)

Comment (qunabu) [reply docs/v3.x/guides/unit-testing.md]: @alexandrebodin - changed, the guide is updated

File: docs/v3.x/guides/unit-testing.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  return instance;
+}
+module.exports = { setupStrapi };
+```
+
+### Test strapi instance
+
+We need a main entry file for our tests, one that will also test our helper file.
+
+**Path —** `./tests/helpers/strapi.js`
```
Comment (iicdii) [inline docs/v3.x/guides/unit-testing.md:131]: I think this needs to be changed to `./tests/app.test.js`

Comment (qunabu) [reply docs/v3.x/guides/unit-testing.md:131]:
```suggestion
**Path —** `./tests/app.test.js`
```

File: docs/v3.x/guides/unit-testing.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+      "/node_modules/",
+      ".tmp",
+      ".cache"
+    ],
+    "testEnvironment": "node"
+  }
+```
+
+Those will inform `Jest` not to look for test inside the folder where it shouldn't.
+::::
```
Comment (lauriejim) [inline docs/v3.x/guides/unit-testing.md:58]: This end the tab.

But this should be used just after the `npm` command.

Few lines before.

File: docs/v3.x/guides/unit-testing.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+:::: tabs
+
+::: tab yarn
+`yarn add jest supertest`
+:::
+
+::: tab npm
+`npm install jest supertest`
+:::
+
```
Comment (qunabu) [inline docs/v3.x/guides/unit-testing.md:29]:
```suggestion
::::

```

File: docs/v3.x/guides/unit-testing.md (hunk 3 above)
Comment (qunabu) [inline docs/v3.x/guides/unit-testing.md:58]:
```suggestion
```

Comment (qunabu) [reply docs/v3.x/guides/unit-testing.md]: Fixed now

**Comment by sa-ma** [comment]
I'm a getting a 403 status when trying to test a Public endpoint

**Comment by VivianSolide** [comment]
> I'm a getting a 403 status when trying to test a Public endpoint

@sa-ma : did you get rid off that?

**Comment by qunabu** [comment]
> > I'm a getting a 403 status when trying to test a Public endpoint
> 
> @sa-ma : did you get rid off that?

Seems you're missing saving to `db` which routes are public. Tests are using different (empty) database. 

I've written and post about this [Allowing roles for test routes](https://medium.com/qunabu-interactive/strapi-testing-more-examples-b09b07088353)

**Comment by VivianSolide** [comment]
> @VivianSolide
> 
> > > I'm a getting a 403 status when trying to test a Public endpoint
> > 
> > 
> > @sa-ma : did you get rid off that?
> 
> Seems you're missing saving to `db` which routes are public. Tests are using different (empty) database.
> 
> I've written and post about this [Allowing roles for test routes](https://medium.com/qunabu-interactive/strapi-testing-more-examples-b09b07088353)

I stumbled upon your articles, very useful thanks @qunabu :pray:

**Pull request merged**